### DISCUSSION TITLE
Should we Trust `X-GitHub-Delivery` for Replay Attack Prevention

### DISCUSSION THREAD

**Discussion by Kynson** [community/community#136297] [Apps, API and Webhooks] [answered]
### Select Topic Area

Question

### Body

I was reading the [best practice for webhooks](https://docs.github.com/en/webhooks/using-webhooks/best-practices-for-using-webhooks) and the last section suggest that we should ensure `X-GitHub-Delivery` header is unique in order to protect the webhook endpoint from replay attacks. 

If I understand correctly, only the payload (request body) is signed by the webhook secret. If attackers would like to replay an event, they can simply alter the header and resend the request. In this case, the header will still be unique. Isn't that the `X-GitHub-Delivery` header fails to serve its purpose? Or I have missed anything?

**Comment by gabrielfrimodig** [comment, marked as answer]
The `X-GitHub-Delivery` header is designed to provide a unique ID for each webhook event that GitHub sends to your endpoint. It is mainly there to help you track individual deliveries, debug issues and things in that nature. It is not intended to be a primary defense against replay attacks. Relying on `X-GitHub-Delivery` will not stop an attack since the header is not included in the HMAC signature that GitHub generates for the payload.

**Reply by jimcarreer** [reply]
Interesting, so I too came to this conclusion; however I can't help to notice a lot of AIs / LLMs recommending this as "replay" protection.  While I'm on the subject; it seems relatively trivial to include the ID in the payload body itself, in which case replay protection could be achieved:

1. Check header first, reject those already handled
2. Validate signature of payload
3. Deserialize and check header against ID in payload, reject those that don't match. 

This would allow for some pretty reasonable replay protection without exposing ingress to deserializing un-validated / un-trusted data.
